Add ByteBufFormat option to LoggingHandler

Bennett-Lynch opened the pull request:
**Motivation**

`LoggingHandler` is a very useful tool for debugging and for tracking the
sequence of events in a pipeline. `LoggingHandler` also includes the
functionality to log a hex dump of all written and received `ByteBuf`s.
This can be useful for small messages, but for large messages, this can
potentially result in extremely large logs. E.g., a 1 MB payload will
result in over a 1 MB log message being recorded. While `LoggingHandler`
may only be intended for debugging, this can still be too excessive in
some debugging scenarios.

**Modifications**

* Create a new `ByteBufFormat` enum that allows users to specify `SIMPLE`
or `HEX_DUMP` logging for `ByteBuf`s.
* For all constructors that currently accept a `LogLevel` parameter,
create new overloaded constructors that also accept this enum as a
parameter.
* Continue to record hex dumps by default.

**Result**

Users will be able to opt out of full hex dump recording, if they wish
to.

Bennett-Lynch commented:
I'm open to the idea of converting the new parameter from an `enum` to `boolean`, if desired. I tend to prefer `enum`s, but I don't see them used too widely in other Netty constructors.

normanmaurer commented:
@trustin @njhill any thoughts / concerns ?

njhill commented on handler/src/main/java/io/netty/handler/logging/ByteBufFormat.java:
```diff
@@ -0,0 +1,20 @@
+ *
+ * @see LoggingHandler
+ */
+public enum ByteBufFormat {
```
Move inside `LoggingHandler`?

normanmaurer replied:
I think it is fine to keep it outside as `LogLevel` is also top-level

njhill commented on handler/src/main/java/io/netty/handler/logging/LoggingHandler.java:
```diff
@@ -317,8 +362,11 @@ private static String formatByteBuf(ChannelHandlerContext ctx, String eventName,
             int rows = length / 16 + (length % 15 == 0? 0 : 1) + 4;
             StringBuilder buf = new StringBuilder(chStr.length() + 1 + eventName.length() + 2 + 10 + 1 + 2 + rows * 80);
```
Also adjust `StringBuilder` capacity accordingly? (same for `ByteBufHolder` case below)

and kind of separate to this change but I wonder if it would be better to use `InternalThreadLocalMap#stringBuilder()` in these methods (maybe we don't care much about the allocations though anyhow in cases where this amount of logging is enabled)

Bennett-Lynch replied:
Will adjust `StringBuilder` capacity, but I'm not sure I fully follow the math. In the sample `shouldLogByteBufDataRead()` unit test, it projects 461 capacity but results in a length of 406. Is it meant to over-estimate or was `appendPrettyHexDump()` updated without updating this logic?

normanmaurer commented on handler/src/main/java/io/netty/handler/logging/ByteBufFormat.java:
```diff
@@ -0,0 +1,20 @@
+package io.netty.handler.logging;
```
@Bennett-Lynch add missing copyright header.

Bennett-Lynch replied:
Will add.

normanmaurer commented on handler/src/main/java/io/netty/handler/logging/ByteBufFormat.java:
```diff
@@ -0,0 +1,20 @@
+import io.netty.buffer.ByteBufUtil;
+
+/**
+ * Used to control the format and verbosity of {@link ByteBuf} logging.
```
Isn't it also for `ByteBufHolder` logging ?

Bennett-Lynch replied:
Will fix.

normanmaurer commented on handler/src/main/java/io/netty/handler/logging/ByteBufFormat.java:
```diff
@@ -0,0 +1,20 @@
+     * {@link ByteBuf}s will be logged using {@link ByteBufUtil#appendPrettyHexDump(StringBuilder, ByteBuf)}.
+     */
+    HEX_DUMP
+}
```
add CRLF for end-of-file

normanmaurer commented on handler/src/main/java/io/netty/handler/logging/LoggingHandler.java:
```diff
@@ -35,18 +35,20 @@
 public class LoggingHandler extends ChannelDuplexHandler {
 
     private static final LogLevel DEFAULT_LEVEL = LogLevel.DEBUG;
+    private static final ByteBufFormat DEFAULT_BYTE_BUF_FORMAT = ByteBufFormat.HEX_DUMP;
```
no need for this static field imho

Bennett-Lynch replied:
Will inline.

normanmaurer commented on handler/src/test/java/io/netty/handler/logging/LoggingHandlerTest.java:
```diff
@@ -270,18 +284,29 @@ public void shouldLogChannelReadComplete() throws Exception {
+            if (actualMsg.matches(expected)) {
+                // The presence of a newline implies a hex-dump was logged
+                return actual.getMessage().contains(NEWLINE) == shouldContainNewline; 
+            } else {
```
nit: remove `else` as you return in the `if` block anyway.

Bennett-Lynch replied:
I like the alternate branches to have the same indentation, but will remove.

normanmaurer commented:
@Bennett-Lynch ping

Bennett-Lynch commented:
> @Bennett-Lynch ping

Thanks for the ping. Addressed feedback.

There was some existing duplicate code between `formatByteBuf()` and `formatByteBufHolder()`. I have opted not to refactor it for now, but can if needed.

normanmaurer commented:
@netty-bot test this please

normanmaurer commented:
@Bennett-Lynch thanks a lot 👍
